src/node/txorphanage.cpp

```diff
@@ -8,68 +8,201 @@
+        boost::multi_index::ordered_unique<boost::multi_index::tag<ByPeer>, ByPeerViewExtractor>
+    >{};
+
+    using AnnouncementMap = boost::multi_index::multi_index_container<Announcement, OrphanIndices>;
+    template<typename Tag>
+    using Iter = typename AnnouncementMap::index<Tag>::type::iterator;
+    AnnouncementMap m_orphans;
+
+    const TxOrphanage::Count m_max_global_latency_score{DEFAULT_MAX_ORPHANAGE_LATENCY_SCORE};
+    const TxOrphanage::Usage m_reserved_usage_per_peer{DEFAULT_RESERVED_ORPHAN_WEIGHT_PER_PEER};
+
+    /** Number of unique orphans by wtxid. Less than or equal to the number of entries in m_orphans. */
+    TxOrphanage::Count m_unique_orphans{0};
+
+    /** Memory used by orphans (see Announcement::GetMemUsage()), deduplicated by wtxid. */
+    TxOrphanage::Usage m_unique_orphan_bytes{0};
+
+    /** The sum of each unique transaction's latency scores including the inputs only (see Announcement::GetLatencyScore
+     * but subtract 1 for the announcements themselves). The total orphanage's latency score is given by this value +
+     * the number of entries in m_orphans. */
+    TxOrphanage::Count m_unique_rounded_input_scores{0};
 
-    /** Orphan transactions in vector for quick random eviction */
-    std::vector<OrphanMap::iterator> m_orphan_list;
+    /** Index from the parents' outputs to wtxids that exist in m_orphans. Used to find children of
+     * a transaction that can be reconsidered and to remove entries that conflict with a block.*/
+    std::unordered_map<COutPoint, std::set<Wtxid>, SaltedOutpointHasher> m_outpoint_to_orphan_it;
 
-    /** Timestamp for the next scheduled sweep of expired orphans */
-    NodeSeconds m_next_sweep{0s};
+    struct PeerDoSInfo {
+        TxOrphanage::Usage m_total_usage{0};
+        TxOrphanage::Count m_count_announcements{0};
+        TxOrphanage::Count m_total_latency_score{0};
+        bool operator==(const PeerDoSInfo& other) const
+        {
+            return m_total_usage == other.m_total_usage &&
+                   m_count_announcements == other.m_count_announcements &&
+                   m_total_latency_score == other.m_total_latency_score;
+        }
+        void Add(const Announcement& ann)
+        {
+            m_total_usage += ann.GetMemUsage();
+            m_total_latency_score += ann.GetLatencyScore();
+            m_count_announcements += 1;
+        }
+        bool Subtract(const Announcement& ann)
+        {
+            Assume(m_total_usage >= ann.GetMemUsage());
+            Assume(m_total_latency_score >= ann.GetLatencyScore());
+            Assume(m_count_announcements >= 1);
+
+            m_total_usage -= ann.GetMemUsage();
+            m_total_latency_score -= ann.GetLatencyScore();
+            m_count_announcements -= 1;
+            return m_count_announcements == 0;
+        }
+        /** There are 2 DoS scores:
+        * - CPU score (ratio of total latency score / max allowed latency score)
```

Comment:
in d5097d6eef85eab971352cf78fb94a30c6e6f127: nit: should this also called "latency score" here for consistency, or is the different naming intentional?